text/0012-milestone-merkle-validation/0012-milestone-merkle-validation.md

```diff
@@ -0,0 +1,128 @@
+
+Where previously the structure of the Tangle alone was sufficient to address those issues, this RFC proposes to add the Merkle tree hash of all the valid (i.e. not ignored) newly confirmed bundles to the signed part of a milestone. This way, each IOTA node can check that the hash matches its local ledger state changes or provide a Merkle audit path for that milestone to prove the inclusion of a particular bundle.
+
+# Detailed design
+
+## Creating a Milestone
+
+- Perform tip selection to choose a branch and a trunk for the new milestone.
+- Determine the topological order according to [RFC-0005](https://github.com/iotaledger/protocol-rfcs/blob/master/text/0005-white-flag/0005-white-flag.md) of the referenced bundles that are not yet confirmed by a previous milestone.
+- Construct the list B<sup>tri</sup> consisting of the tail transaction hashes of all the not-ignored bundles in that particular order.
+- Convert each element of B<sup>tri</sup> to binary by splitting it into groups of 5 trits and interpreting each group as a balanced ternary value in little-endian representation. Each value is then encoded as a signed (two's complement) 8-bit integer. (This exactly matches the conversion used for binary I/O of ternary data in the current protocol.) This leads to the ordered list B containing 49-byte strings.
+- Compute the 64-byte Merkle tree hash H = MTH(B).
+- Encode H into ternary by interpreting each octet of the string H as a signed 8-bit integer value v and then encoding v as a little-endian 6-trit string in balanced ternary representation. This leads to H<sup>tri</sup> with a length of 384 trits.
+- Prepare the milestone bundle as usual. Its head transaction contains the information required to verify the Coordinator's signature in its `signatureMessageFragment` field. This information has a length of d·81 trytes, where d is the depth of the Coordinator's Merkle tree. 
+- Append H<sup>tri</sup> to the `signatureMessageFragment` field. For any depth d < 26 the field provides sufficient space.
+- Sign the head transaction and add its fragmented signature to the milestone bundle's zero value transactions.
+
+## Milestone validation
+
+- Verify the signature of the milestone m.
+- Construct the ordered list B<sup>tri</sup> of the tail transaction hashes of the not-ignored bundles m confirms.
+- Encode the hashes B<sup>tri</sup> into their binary representation B and compute H = MTH(B).
+- Extract the first 192 trits after the Coordinator's Merkle tree information from the `signatureMessageFragment` field of the head transaction and verify that this matches the ternary encoded H.
```

Comment:
Above H<sup>tri</sup> is 384 trits (stemming from encoding 6trits*64) in length but here it is only 192 trits, is that correct?

Reply:
No, you are absolutely right, since we are using 512-bit hash it will be 284 trits. Will fix that

Reply:
fixed